protoc-gen-gogo/generator/generator.go

```diff
@@ -2332,13 +2332,29 @@ func (g *Generator) generateMessage(message *Descriptor) {
 	// Wrapper for table-driven marshaling and unmarshaling.
 	g.P("func (m *", ccTypeName, ") XXX_Unmarshal(b []byte) error {")
 	g.In()
-	g.P("return xxx_messageInfo_", ccTypeName, ".Unmarshal(m, b)")
+	if gogoproto.IsUnmarshaler(g.file.FileDescriptorProto, message.DescriptorProto) {
+		g.P("return m.Unmarshal(b)")
+	} else {
+		g.P("return xxx_messageInfo_", ccTypeName, ".Unmarshal(m, b)")
+	}
 	g.Out()
 	g.P("}")
 
 	g.P("func (m *", ccTypeName, ") XXX_Marshal(b []byte, deterministic bool) ([]byte, error) {")
 	g.In()
-	g.P("return xxx_messageInfo_", ccTypeName, ".Marshal(b, m, deterministic)")
+	if gogoproto.IsMarshaler(g.file.FileDescriptorProto, message.DescriptorProto) ||
+		gogoproto.IsUnsafeMarshaler(g.file.FileDescriptorProto, message.DescriptorProto) {
```

Comment:
I currently left out the IsStableMarshaler check. Is that fine? Are we saying the functionality is correct and the documentation must be updated?